ecma402_traits/src/listformat.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// Contains the API configuration as prescribed by ECMA 402.
+///
+/// The meaning of the options is the same as in the similarly named
+/// options in the JS version.
+///
+/// See [Options] for the contents of the options.  See the [Format::try_new]
+/// for the use of the options.
+pub mod options {
+    /// Chooses the list formatting approach.
+    #[derive(Eq, PartialEq, Debug, Clone)]
+    pub enum Style {
+        Long,
+        Short,
+        Narrow,
+    }
+    /// Chooses between "this, that and other", and "this, that or other".
+    #[derive(Eq, PartialEq, Debug, Clone)]
+    pub enum Type {
+        /// "This, that and other".
+        Conjunction,
+        /// "This, that or other".
+        Disjunction,
+    }
+}
+
+/// The options set by the user at construction time.  See discussion at the top level
+/// about the name choice.  Provides as a "bag of options" since we don't expect any
+/// implementations to be attached to this struct.
+///
+/// The default values of all the options are prescribed in by the [TC39 report][tc39lf].
+///
+///   [tc39lf]: https://tc39.es/proposal-intl-list-format/#sec-Intl.ListFormat
+#[derive(Eq, PartialEq, Debug, Clone)]
+pub struct Options {
+    /// Selects a [options::Style] for the formatted list.  If unset, defaults
+    /// to [options::Style::Long].
+    pub style: options::Style,
+    /// Selects a [options::Type] for the formatted list.  If unset, defaults to
+    /// [options::Type::Conjunction].
+    pub in_type: options::Type,
+}
+
+/// Allows the use of `listformat::Format::try_new(..., Default::default())`.
+impl Default for Options {
+    /// Gets the default values of [Options] if omitted at setup.  The
+    /// default values are prescribed in by the [TC39 report][tc39lf].
+    ///
+    ///   [tc39lf]: https://tc39.es/proposal-intl-list-format/#sec-Intl.ListFormat
+    fn default() -> Self {
+        Options {
+            style: options::Style::Long,
+            in_type: options::Type::Conjunction,
+        }
+    }
+}
+
+use std::fmt;
+
+/// The package workhorse: formats supplied pieces of text into an ergonomically formatted
+/// list.
+///
+/// While ECMA 402 originally has functions under `Intl`, we probably want to
+/// obtain a separate factory from each implementor.
+///
+/// Purposely omitted:
+///
+/// - `supported_locales_of`.
+pub trait Format {
```

Comment:
I would name it `ListFormatType`.

Reply:
Hey @zbraniecki,thank you for taking a look.

Would you mind giving some detail on the naming approach?  I personally dislike noisy names.  In case of `listformat::ListFormatType`, IMHO the phrase `listformat` is repetitive, and `Type` doesn't clarify the type, but adds more text to its name.  But that's just my personal opinion and I would gladly adopt another approach if I understood the rationale.  There's even a note on that in [rust language style[rls].

[rls]: https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#avoid-redundant-prefixes-[rfc-356]

Now since you've written *way* more Rust code than me, you probably know better what has worked well for you in the past. That's why I ask.

Reply:
Since you lgtm'd the code as is now, I'll go on and commit.  This is not being released anywhere so I think it's OK to put a milestone in by committing.  It's a tall stack of changes as it is, and I'd be glad to make touch ups based on followup agreements.

Reply:
Yeah, totally. Land at will.

> Would you mind giving some detail on the naming approach?

Structs are often imported to be used in code without module namespacing. The example you provided from RLS (from 2013!) makes sense for more internal code:

> Names of items within a module should not be prefixed with that module's name:

So, when designing more internal names, I'm ok with `Options`, `Error`, `Format` etc. But public names, ones used across crates, should be unique.

Eventually, I believe we'll want someone to write:

```rust
let dtf = DateTimeFormat::new();
let lf = ListFormat::new();
let nf = NumberFormat::new();
```

rather than:

```rust
let dtf = datetimeformat::Format::new();
let lf = listformat::Format::new();
let nf = numberformat::Format::new();
```

If your public API doesn't allow for that, then users will have to do:

```rust
use icu_datetime::Format as DateTimeFormat;
use icu_listformat::Format as ListFormat;
use icu_numberformat::Foramt as NumberFormat;

let dtf = DateTimeFormat::new();
let lf = ListFormat::new();
let nf = NumberFormat::new();
```

Now, saying that, since `Format` is unique inside the crate, if you really prefer, I'm ok with keeping the name `Format` internally, and in `src/lib.rs` doing 

```rust
pub use Format as ListFormat;
pub use options::Options as ListFormatOptions;
```

but I think it may defeat the purpose.

@Manishearth - do you have any thoughts on naming scheme? It may be just my personal bias.

Reply:
Got it. 

I think the main takeaway is that if users expect to import each type by name, then we might as well have names that reflect this expectation.  If anything, then it's going to be less surprising.  Regardless, there's something to be said about meeting expectations and the principle of least surprise.  I'll rework the naming to reflect this understanding.  I'll also wait for a while for an opinion from Manish, but wanted to note early that I (think I) understand your point. (I also understand that naming is a bit of a bikeshed, and I'd rather use *any* naming schema than spend more than a minimal amount of time discussing naming :) Hence, let's agree early and proceed.)

I have an impression that this approach to naming is a bit of a C++-ism (citation_needed).  Not to say that it's bad, it's just that the coloring shows through. Namespaces were introduced fairly late (measured from the time we had the need to integrate libraries with naming conflicts!), so we already had to encode disambiguations in type names.  And when namespaces become widely available we just prepended them to type names, so that qualified names were never fully utilized.  For example, I would personally rather have `listformat::Type` than `listformat::ListFormatType`, but just as before, I'm not dogmatic on this.